src/Razor/test/Microsoft.AspNetCore.Razor.ProjectEngineHost.Test/Utilities/UrlDecoderTests.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Test.Common;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.AspNetCore.Razor.ProjectEngineHost.Test.Utilities;
+
+public class UrlDecoderTests(ITestOutputHelper testOutput) : TestBase(testOutput)
+{
+    public static IEnumerable<object[]> UrlDecodeData =>
+        new[]
+        {
+            new object[] { "http://127.0.0.1:8080/appDir/page.aspx?foo=bar", "http://127.0.0.1:8080/appDir/page.aspx?foo=b%61r" },
+            new object[] { "http://127.0.0.1:8080/appDir/page.aspx?foo=b%ar", "http://127.0.0.1:8080/appDir/page.aspx?foo=b%%61r" },
+            new object[] { "http://127.0.0.1:8080/app%Dir/page.aspx?foo=b%ar", "http://127.0.0.1:8080/app%Dir/page.aspx?foo=b%%61r" },
+            new object[] { "http://127.0.0.1:8080/app%%Dir/page.aspx?foo=b%%r", "http://127.0.0.1:8080/app%%Dir/page.aspx?foo=b%%r" },
+            new object[] { "http://127.0.0.1:8080/appDir/page.aspx?foo=ba%r", "http://127.0.0.1:8080/appDir/page.aspx?foo=b%61%r" },
+            new object[] { "http://127.0.0.1:8080/appDir/page.aspx?foo=bar baz", "http://127.0.0.1:8080/appDir/page.aspx?foo=bar+baz" },
+            new object[] { "http://example.net/\U00010000", "http://example.net/\U00010000" },
+            new object[] { "http://example.net/\uD800", "http://example.net/\uD800" },
+            new object[] { "http://example.net/\uD800a", "http://example.net/\uD800a" },
+            // The "Baz" portion of "http://example.net/Baz" has been double-encoded - one iteration of UrlDecode() should produce a once-encoded string.
+            new object[] { "http://example.net/%42%61%7A", "http://example.net/%2542%2561%257A"},
```

Comment:
If you happen to be doing another rev, a test that proves that lowercase hex chars are accepted in addition to uppercase might be a nice case.